### PR TITLE
(proof-new) Theory engine proof producing

### PR DESCRIPTION
**Pull request by ajreynol**
This completes proof support in TheoryEngine. 

The main addition in this PR is to make its getExplanation method proof producing.

### CODE REVIEW

File: src/theory/theory_engine.cpp [hunk 1]
```diff
@@ -1759,10 +1822,139 @@ theory::TrustNode TheoryEngine::getExplanation(
...
+      if (tExp == tConc)
+      {
+        // trivial
+        Trace("te-proof-exp") << "...trivial" << std::endl;
+        continue;
+      }
+      // ------------- Via theory
+      // tExp => tConc              tExp
+      // ---------------------------------MACRO_SR_PRED_TRANSFORM
+      // tConc
```
Comment (HanielB) [inline src/theory/theory_engine.cpp:1931]:
```suggestion
      // ------------- Via theory
      // tExp => tConc              tExp
      // -------------IMPLIES_ELIM
      // ~tExp v tConc              tExp
      // ---------------------------------RESOLUTION_{tExp}
      // tConc
```

Feels a bit too much to use MACRO_SR_PRED_TRANSFORM here. :) Or is there something else going on?

File: src/theory/theory_engine.cpp [hunk 2]
```diff
@@ -1759,10 +1822,139 @@ theory::TrustNode TheoryEngine::getExplanation(
...
+        // otherwise, trusted theory lemma
+        Node tidn = builtin::BuiltinProofRuleChecker::mkTheoryIdNode(it->first);
+        lcp->addStep(proven, PfRule::THEORY_LEMMA, {}, {proven, tidn});
+      }
+      std::vector<Node> pfChildren;
+      pfChildren.push_back(proven);
+      pfChildren.push_back(trn.getNode());
+      std::vector<Node> pfArgs;
+      pfArgs.push_back(tConc);
+      pfArgs.push_back(mkMethodId(MethodId::SB_FORMULA));
```
Comment (HanielB) [inline src/theory/theory_engine.cpp:1949]: why is this necessary?

File: src/theory/theory_engine.cpp [hunk 3]
```diff
@@ -947,18 +954,16 @@ void TheoryEngine::assertToTheory(TNode assertion, TNode originalAssertion, theo
...
-    Assert(atom.getKind() == kind::EQUAL)
-        << "atom should be an EQUALity, not `" << atom << "'";
+    Assert(assertion.getKind() == kind::EQUAL
+           || (assertion.getKind() == kind::NOT
+               && assertion[0].getKind() == kind::EQUAL))
+        << "atom should be an EQUALity, not `" << assertion << "'";
     if (markPropagation(assertion, originalAssertion, toTheoryId, fromTheoryId)) {
       // assert to the shared solver
+      bool polarity = assertion.getKind() != kind::NOT;
+      TNode atom = polarity ? assertion : assertion[0];
```
[the remark this answers is not in the document]
Comment (ajreynol) [reply src/theory/theory_engine.cpp:966]: Polarity is passed to `assertSharedEquality` below.

File: src/theory/theory_engine.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/theory_engine.cpp:1930]: No, this was written before the resolution rule existed :)

I'll update.

File: src/theory/theory_engine.cpp (hunk 2 above)
Comment (ajreynol) [reply src/theory/theory_engine.cpp:1948]: This made it so that tExp is applied a formula substitution, e.g. `tExp -> true` instead of being interpreted potential as something else.  If this is replaced by resolution this is irrelevant.

File: src/theory/theory_engine.cpp (hunk 1 above)
Comment (HanielB) [reply src/theory/theory_engine.cpp:1930]: Cool

**Review by HanielB — APPROVED** [review]

File: src/theory/theory_engine.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/theory_engine.cpp:1930]: Will address on later PR.  We decided on a MODUS_PONENS rule.
